speakerbox/main.py

```diff
@@ -0,0 +1,202 @@
+        Default: 2
+
+    Returns
+    -------
+    model_storage_path: Path
+        The path to the directory where the model is stored.
+    """
+    # Load feature extractor
+    feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(model_base)
+
+    # Convert provided dataset to HF Dataset
+    if isinstance(dataset, pd.DataFrame):
+        dataset = Dataset.from_pandas(dataset, preserve_index=False)
+        dataset = dataset.class_encode_column("label")
+        dataset = dataset.cast_column("audio", Audio(feature_extractor.sampling_rate))
+
+    # Convert into train, test, and validate dict
+    if isinstance(dataset, Dataset):
+        train_and_test = dataset.train_test_split(test_size=0.4)
+        test_and_valid = train_and_test["test"].train_test_split(test_size=0.5)
+        ds_dict = DatasetDict(
+            {
+                "train": train_and_test["train"],
+                "test": test_and_valid["train"],
+                "valid": test_and_valid["test"],
+            }
+        )
+
+    # Show dataset summary stats
+    for subset in ["train", "test", "valid"]:
+        pd_subset = ds_dict[subset].to_pandas()
+        print(f"Summary stats for '{subset}' dataset")
+        print(f"n-labels: {pd_subset.label.nunique()}")
+        print(f"Avg duration: {pd_subset.duration.mean()}")
+        print(f"Min duration: {pd_subset.duration.min()}")
+        print(f"Max duration: {pd_subset.duration.max()}")
+        print(f"StD duration: {pd_subset.duration.std()}")
+        print("-" * 80)
+
+    # Construct label to id and vice-versa LUTs
+    label2id, id2label = dict(), dict()
+    for i, label in enumerate(ds_dict["train"].features["label"].names):
+        label2id[label] = str(i)
+        id2label[str(i)] = label
+
+    # Construct preprocessing function
+    def preprocess(
+        examples: arrow_dataset.Batch,
+    ) -> feature_extraction_utils.BatchFeature:
+        audio_arrays = [x["array"] for x in examples["audio"]]
+        inputs = feature_extractor(
+            audio_arrays,
+            sampling_rate=feature_extractor.sampling_rate,
+            max_length=int(feature_extractor.sampling_rate * max_duration),
+            do_normalize=True,
+            truncation=True,
+            padding=True,
+        )
+        return inputs
+
+    # Encode the dataset
+    ds_dict = ds_dict.map(preprocess, batched=True)
+
+    # Create AutoModel
+    model = Wav2Vec2ForSequenceClassification.from_pretrained(
+        model_base,
+        num_labels=len(id2label),
+        label2id=label2id,
+        id2label=id2label,
+        ignore_mismatched_sizes=True,
+    )
+
+    # Create fine tuning Trainer
+    args = TrainingArguments(
+        model_name,
+        evaluation_strategy="epoch",
+        save_strategy="epoch",
+        learning_rate=3e-5,
+        per_device_train_batch_size=batch_size,
+        gradient_accumulation_steps=4,
+        per_device_eval_batch_size=batch_size,
+        num_train_epochs=5,
```

Comment:
I think time and money could be saved if the training could be stopped early, instead of going all the way to a fixed epoch. When the training acc keeps improving, but the test acc plateaus or goes lower, overfitting has occurred and training can be stopped.

Reply:
Totally agree on the overfitting but the time and money is really low... on my 1080TI training takes all of 15 minutes. That said, will look into `Trainer` / `TrainingArguments` options for stopping training after some loss threshold.

Reply:
Thanks for looking into it. I'm worried others might not have a GPU available and would have to use AWS or something else.

Reply:
looks like I can add the following to `Trainer`:

```python
callbacks=[
    EarlyStoppingCallback(
        early_stopping_patience=2,  # num evals to try with threshold until exit
        early_stopping_threshold=0.02,  # acc must improve my this much or patience decr
    )
],
```

Reply:
Going to add that to Trainer _after_ I test it on a random transcript / get more data though imo